serialize/src/lib.rs

```diff
@@ -94,6 +96,37 @@ pub trait CanonicalSerialize {
     }
 }
 
+// This private struct works around Serialize taking the pre-existing
+// std::io::Write instance of most digest::Digest implementations by value
+struct HashMarshaller<'a, H: Digest>(&'a mut H);
+
+impl<'a, H: Digest> ark_std::io::Write for HashMarshaller<'a, H> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> ark_std::io::Result<usize> {
+        Digest::update(self.0, buf);
+        Ok(buf.len())
+    }
+
+    #[inline]
+    fn flush(&mut self) -> ark_std::io::Result<()> {
+        Ok(())
+    }
+}
+
+/// The CanonicalSerialize induces a natural way to hash the
+/// corresponding value, of which this is the convenience trait.
+pub trait Hash: CanonicalSerialize {
+    fn hash<H: Digest>(&self) -> GenericArray<u8, <H as Digest>::OutputSize> {
+        let mut hasher = H::new();
+        self.serialize(HashMarshaller(&mut hasher))
+            .expect("HashMarshaller::flush should be infaillible!");
+        hasher.finalize()
+    }
+}
+
+/// Hash is a (blanket) extension trait of CanonicalSerialize
+impl<T: CanonicalSerialize> Hash for T {}
```

Comment:
Should we just make this a standalone method(s)?
```rust
fn hash_with_digest_compressed<D: Digest, T: CanonicalSerialize>() -> GenericArray<u8, <H as Digest>::OutputSize> { ... }

fn hash_with_digest_uncompressed<D: Digest, T: CanonicalSerialize>() -> GenericArray<u8, <H as Digest>::OutputSize> { ... }
```

Not sure of the trade-off between importing a trait (`Hash`) vs. two methods. What do you think?

Regardless, I think we should rename it from `Hash` to something else, as it collides with `core::hash::Hash`.

Reply:
> Not sure of the trade-off between importing a trait (Hash) vs. two methods.

The [extension trait pattern](https://rust-lang.github.io/rfcs/0445-extension-trait-conventions.html) is clear enough that most will prefer it to manually importing several methods, I think.

I renamed the trait according to best practices.

Reply:
I could also simply add those two methods to the `CanonicalSerialize` trait? Wdyt?

Reply:
I prefer the separation of concerns, and maybe we should put this behind a separate feature?